src/builder.rs

```diff
@@ -0,0 +1,831 @@
+    ) -> &mut Self {
+        self.block_owner_deletion = block_owner_deletion.into();
+        self
+    }
+
+    pub fn controller<VALUE: Into<bool>>(&mut self, controller: VALUE) -> &mut Self {
+        self.controller = Some(controller.into());
+        self
+    }
+
+    pub fn controller_opt<VALUE: Into<Option<bool>>>(&mut self, controller: VALUE) -> &mut Self {
+        self.controller = controller.into();
+        self
+    }
+
+    pub fn kind<VALUE: Into<String>>(&mut self, kind: VALUE) -> &mut Self {
+        self.kind = Some(kind.into());
+        self
+    }
+
+    pub fn kind_opt<VALUE: Into<Option<String>>>(&mut self, kind: VALUE) -> &mut Self {
+        self.kind = kind.into();
+        self
+    }
+
+    pub fn name<VALUE: Into<String>>(&mut self, name: VALUE) -> &mut Self {
+        self.name = Some(name.into());
+        self
+    }
+
+    pub fn name_opt<VALUE: Into<Option<String>>>(&mut self, name: VALUE) -> &mut Self {
+        self.name = name.into();
+        self
+    }
+
+    pub fn uid<VALUE: Into<String>>(&mut self, uid: VALUE) -> &mut Self {
+        self.uid = Some(uid.into());
+        self
+    }
+
+    pub fn uid_opt<VALUE: Into<Option<String>>>(&mut self, uid: VALUE) -> &mut Self {
+        self.uid = uid.into();
+        self
+    }
+
+    /// Can be used to initialize a builder with settings from an existing resource.
+    /// The builder will create an `OwnerReference` that points to the passed resource.
+    ///
+    /// This will _not_ set `controller` or `block_owner_deletion`.
+    pub fn initialize_from_resource<T: Resource<DynamicType = ()>>(
+        &mut self,
+        resource: &T,
+    ) -> &mut Self {
+        self.api_version(T::api_version(&()))
+            .kind(T::kind(&()))
+            .name(resource.name())
+            .uid_opt(resource.meta().uid.clone());
+        self
+    }
+
+    pub fn build(&self) -> OperatorResult<OwnerReference> {
+        Ok(OwnerReference {
+            api_version: match self.api_version {
+                None => return Err(Error::MissingObjectKey { key: "api_version" }),
+                Some(ref api_version) => api_version.clone(),
+            },
+            block_owner_deletion: self.block_owner_deletion,
+            controller: self.controller,
+            kind: match self.kind {
+                None => return Err(Error::MissingObjectKey { key: "kind" }),
+                Some(ref kind) => kind.clone(),
+            },
+            name: match self.name {
+                None => return Err(Error::MissingObjectKey { key: "name" }),
+                Some(ref name) => name.clone(),
+            },
+            uid: match self.uid {
+                None => return Err(Error::MissingObjectKey { key: "uid" }),
+                Some(ref uid) => uid.clone(),
+            },
+        })
+    }
+}
+
+/// A builder to build [`Pod`] objects.
+///
+#[derive(Clone, Default)]
+pub struct PodBuilder {
+    metadata: Option<ObjectMeta>,
```

Comment:
There are four functions to set metadata and a user can easily overwrite it by mistake.

I think the code would be much cleaner and less error prone if this were an `ObjectMetaBuilder` instead of a "materialized" `ObjectMeta`. The only time this would be materialized would be at `build()` time.

Reply:
I understand your concern I don't think I understand your proposed solution.

Can you elaborate?

Reply:
I wanted to prepare a patch, but I discovered that the patch would be quite big, so I' trying to sketch the idea here:

```
#[derive(Clone, Default)]
pub struct PodBuilder {
    metadata_builder: ObjectMetaBuilder,
// other fields skipped
}

impl PodBuilder {
    pub fn new() -> PodBuilder {
        PodBuilder {
            metadata: ObjectMetaBuilder::new(),
            ..PodBuilder::default()
        }
    }

   // the only method you need to set meta data fields
   // keep on reusing the inner object meta builder
    pub fn metadata<F>(&mut self, f: F) -> &mut Self
    where
        F: Fn(&mut ObjectMetaBuilder) -> &mut ObjectMetaBuilder,
    {
        f(&mut self.metadata_builder);
        self
    }

// other methods skipped

    /// Consumes the Builder and returns a constructed Pod
    pub fn build(&self) -> OperatorResult<Pod> {

// skip some code here ... and at the result of build() is:
        Ok(Pod {
            metadata: match self.metadata.build() {
                Err => return Err(Error::MissingObjectKey { key: "metadata" }),
                Ok(ref metadata) => metadata.clone(),
            },
            spec: Some(PodSpec {
                containers: self.containers.clone(),
                tolerations: self.tolerations.clone(),
                volumes,
                node_name: self.node_name.clone(),
                ..PodSpec::default()
            }),
            status: self.status.clone(),
        })
   }
}
```

Reply:
I added the `metadata_default` so it can be used in tests mor easily.
I think `metadata` makes sense for when you really do want to override it all
`metadata_opt` can go for now I agree

I like your idea of having a reusable builder in theory. In practice I'm not sure how often that'd be used really.